# Answering Business Questions using SQL

## Introduction

We'll be working with a version of a database called Chinook. The Chinook database contains information about a fictional digital music shop - kind of like a mini-iTunes store.

The Chinook database contains information about the artists, songs, and albums from the music shop, as well as information on the shop's employees, customers, and the customers purchases. This information is contained in eleven tables. 

Now we'll connect our Jupyter Notebook to our database file.

In [2]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

## Overview the Data

Let's start by getting familiar with the data. 

In [3]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table'),
 ('name', 'view'),
 ('bobo', 'view'),
 ('bobo1', 'view'),
 ('tablew', 'view'),
 ('tttt', 'view')]

In [4]:
%%sql
SELECT
    *
FROM customer 

Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4),
 (6, 'Helena', 'Holý', None, 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', '+420 2 4177 0449', None, 'hholy@gmail.com', 5),
 (7, 'Astrid', 'Gruber', None, 'Rotenturmstraße 4, 1010 Innere Stadt', 'Vienne', None, 'Austria', '1010', '+43 01 5134505', None, 'astrid.gruber@apple.at', 5),
 (8, 'Daan', 'Peeters', None, 'Grétrystraat 63', 'Brussels', None, 'Belgium', '1000', '+32 02 219 03 03', None, 'daan_peeters@apple.be', 4),
 (9, 'Kara', 'Nielsen', None, 'Sønder Boulevard 51', 'Copenhagen', None, 'Denmark', '1720', '+453 3331 9991', None, 'kara.nielsen@jubii.dk', 4),
 (10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4),
 (11, 'Alexandre', 'Rocha', 'Banco do Brasil S.A.', 'Av. Paulista, 2022', 'São Paulo', 'SP', 'Brazil', '01310-200', '+55 (11) 3055-3278', '+55 (11) 3055-8131', 'alero@uol.com.br', 5),
 (12, 'Roberto', 'Almeida', 'Riotur', 'Praça Pio X, 119', 'Rio de Janeiro', 'RJ', 'Brazil', '20040-020', '+55 (21) 2271-7000', '+55 (21) 2271-7070', 'roberto.almeida@riotur.gov.br', 3),
 (13, 'Fernanda', 'Ramos', None, 'Qe 7 Bloco G', 'Brasília', 'DF', 'Brazil', '71020-677', '+55 (61) 3363-5547', '+55 (61) 3363-7855', 'fernadaramos4@uol.com.br', 4),
 (14, 'Mark', 'Philips', 'Telus', '8210 111 ST NW', 'Edmonton', 'AB', 'Canada', 'T6G 2C7', '+1 (780) 434-4554', '+1 (780) 434-5565', 'mphilips12@shaw.ca', 5),
 (15, 'Jennifer', 'Peterson', 'Rogers Canada', '700 W Pender Street', 'Vancouver', 'BC', 'Canada', 'V6C 1G8', '+1 (604) 688-2255', '+1 (604) 688-8756', 'jenniferp@rogers.ca', 3),
 (16, 'Frank', 'Harris', 'Google Inc.', '1600 Amphitheatre Parkway', 'Mountain View', 'CA', 'USA', '94043-1351', '+1 (650) 253-0000', '+1 (650) 253-0000', 'fharris@google.com', 4),
 (17, 'Jack', 'Smith', 'Microsoft Corporation', '1 Microsoft Way', 'Redmond', 'WA', 'USA', '98052-8300', '+1 (425) 882-8080', '+1 (425) 882-8081', 'jacksmith@microsoft.com', 5),
 (18, 'Michelle', 'Brooks', None, '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', '+1 (212) 221-3546', '+1 (212) 221-4679', 'michelleb@aol.com', 3),
 (19, 'Tim', 'Goyer', 'Apple Inc.', '1 Infinite Loop', 'Cupertino', 'CA', 'USA', '95014', '+1 (408) 996-1010', '+1 (408) 996-1011', 'tgoyer@apple.com', 3),
 (20, 'Dan', 'Miller', None, '541 Del Medio Avenue', 'Mountain View', 'CA', 'USA', '94040-111', '+1 (650) 644-3358', None, 'dmiller@comcast.com', 4),
 (21, 'Kathy', 'Chase', None, '801 W 4th Street', 'Reno', 'NV', 'USA', '89503', '+1 (775) 223-7665', None, 'kachase@hotmail.com', 5),
 (22, 'Heather', 'Leacock', None, '120 S Orange Ave', 'Orlando', 'FL', 'USA', '32801', '+1 (407) 999-7788', None, 'hleacock@gmail.com', 4),
 (23, 'John', 'Gordon', None, '69 Salem Street', 'Boston', 'MA', 'USA', '2113', '+1 (617) 522-1333', None, 'johngordon22@yahoo.com', 4),
 (24, 'Frank', 'Ralston', None, '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', '+1 (312) 332-3232', None, 'fralston@gmail.com', 3),
 (25, 'Victor', 'Stevens', None, '319 N. Frances Street', 'Madison', 'WI', 'USA', '53703', '+1 (608) 257-0597', N

## Selecting Album to Purchase

The Chinook record store has just signed a deal with a new record label, and we've been tasked with selecting the first three albums that will be added to the store, from a list of four. All four albums are by artists that don't have any tracks in the store right now - we have the artist names, and the genre of music they produce:

* *Regal* - *Hip-Hop*
* *Red Tone* - *Punk*
* *Meteor and the Girls* - *Pop*
* *Slim Jim Bites* - *Blues*

The record label specializes in artists from the USA, so we're interested in finding out which genres sell the best in the USA.

Now we'll write a query to find out which genres sell the most tracks in the USA in order to make a recommendation for the three artists whose album should purchase for the store.

In [5]:
%%sql
SELECT
    g.name,
    SUM(il.quantity) tracks_sold,
    ROUND(CAST(SUM(il.quantity) AS Float)*100/(SELECT SUM(quantity)
FROM invoice_line il
LEFT JOIN invoice i ON il.invoice_id=i.invoice_id
WHERE i.billing_country = "USA"),2) || "%" percentage 
FROM genre g
LEFT JOIN track t ON g.genre_id=t.genre_id
LEFT JOIN invoice_line il ON t.track_id=il.track_id
LEFT JOIN invoice i ON i.invoice_id=il.invoice_id
WHERE i.billing_country == "USA"
GROUP BY 1
ORDER BY 2 DESC

Done.


[('Rock', 561, '53.38%'),
 ('Alternative & Punk', 130, '12.37%'),
 ('Metal', 124, '11.8%'),
 ('R&B/Soul', 53, '5.04%'),
 ('Blues', 36, '3.43%'),
 ('Alternative', 35, '3.33%'),
 ('Latin', 22, '2.09%'),
 ('Pop', 22, '2.09%'),
 ('Hip Hop/Rap', 20, '1.9%'),
 ('Jazz', 14, '1.33%'),
 ('Easy Listening', 13, '1.24%'),
 ('Reggae', 6, '0.57%'),
 ('Electronica/Dance', 5, '0.48%'),
 ('Classical', 4, '0.38%'),
 ('Heavy Metal', 3, '0.29%'),
 ('Soundtrack', 2, '0.19%'),
 ('TV Shows', 1, '0.1%')]

Now we can take a decision: the Punk genre is by far the second most popular in the USA (12.37%) so we should purchase the album of the artist *Red Tone* for the store. Purchase the *Slim Jim Bites*' album seems to be a good choice as well since the genre *Blues* is at the fourth place (3.43%).The genres of the last two artists left are very close each other , but we should have a small preference in buying the *Pop* album that seems to be slighty preferable to the *Hip-Hop* one (2.09% vs 1.9%).

Ultimately we suggest the purchase of the album of these three artists :

* *Red Tone* - *Punk*
* *Meteor and the Girls* - *Pop*
* *Slim Jim Bites* - *Blues*

## Analyzing Employee Sales Performance

Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase. We want to analyze the purchases of customers belonging to each employee to see if any sales support agent is performing either better or worse than the others.

In [6]:
%%sql
SELECT
    e.first_name || " " || e.last_name name,
    ROUND(SUM(i.total),2) total,
    SUBSTR(e.hire_date,1,10) hire_date
FROM employee e
LEFT JOIN customer c ON c.support_rep_id=e.employee_id
LEFT JOIN invoice i ON i.customer_id=c.customer_id
WHERE e.title = "Sales Support Agent"
GROUP BY 1
ORDER BY 2 DESC

Done.


[('Jane Peacock', 1731.51, '2017-04-01'),
 ('Margaret Park', 1584.0, '2017-05-03'),
 ('Steve Johnson', 1393.92, '2017-10-17')]

We can note that the sales support agents have similar results until now , and analyzing their hire date we can observe how the total purchase is dependent from the hire date , the longer the agent works the higher the earnings. However seems that Steve Johson have the best result taking into account the fact that he has been hired five or six months or so after the other agents.

## Analyzing Sales by Country

Our next task is to analyze the sales data for customers from each different country. We have been given guidance to use the country value from the customers table, and ignore the country from the billing address in the invoice table.

In particular, we'll calculate data, for each country, on the:

* *total number of customers*
* *total value of sales*
* *average value of sales per customer*
* *average order value*

We'll group the countries with only one customer as "Other".

In [7]:
%%sql
WITH find_other AS
    (
    SELECT
       CASE
           WHEN (
                 SELECT count(*)
                 FROM customer
                 where country = c.country
                ) = 1 THEN "Other"
           ELSE c.country
       END AS country,
       c.customer_id,
       i.*
     FROM invoice i
     INNER JOIN customer c ON c.customer_id = i.customer_id
    )

SELECT 
    country,
    COUNT(DISTINCT(customer_id)) customers_number,
    ROUND(SUM(total),2) total_sales,
    ROUND((SUM(total))/COUNT(DISTINCT(customer_id)),2) avg_per_costumer,
    ROUND(SUM(total)/COUNT(invoice_id),2) avg_order
FROM 
    (
    SELECT
        fo.*,
        CASE
            WHEN fo.country = "Other" THEN 1
            ELSE 0
        END AS sort
    FROM find_other fo
    )
GROUP BY 1
ORDER BY sort , total_sales DESC

Done.


[('USA', 13, 1040.49, 80.04, 7.94),
 ('Canada', 8, 535.59, 66.95, 7.05),
 ('Brazil', 5, 427.68, 85.54, 7.01),
 ('France', 5, 389.07, 77.81, 7.78),
 ('Germany', 4, 334.62, 83.66, 8.16),
 ('Czech Republic', 2, 273.24, 136.62, 9.11),
 ('United Kingdom', 3, 245.52, 81.84, 8.77),
 ('Portugal', 2, 185.13, 92.57, 6.38),
 ('India', 2, 183.15, 91.57, 8.72),
 ('Other', 15, 1094.94, 73.0, 7.45)]

According to the data, there may be good investement opportunities in these countries :

* *Czech Republic*
* *Portugal*
* *India*

Anyway our results are not very reliable due to the low amount of data. We need to collect more data to increase the sample size or at least the company should start with a small investement in these countries. 

## Album vs Individual Tracks

Management are currently considering changing their purchasing strategy to save money. The strategy they are considering is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.

We have been asked to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.

In [21]:
%%sql
WITH invoice_schema AS
    (
    SELECT
        il.invoice_id invoice_id,
        MAX(il.track_id) last_track_id
    FROM invoice_line il
    GROUP BY 1    
    )
    
SELECT
    purchase_type,
    COUNT(invoice_id) invoices_number,
    ROUND(CAST(COUNT(invoice_id) AS Float)*100/(SELECT COUNT(*) FROM invoice),2) percentage
FROM
    (
    SELECT
        i_s.*,
        CASE 
            WHEN 
                (
                SELECT t.track_id FROM track t
                WHERE t.album_id = (
                                    SELECT t2.album_id FROM track t2
                                    WHERE t2.track_id = i_s.last_track_id
                                   )
                EXCEPT
                    
                SELECT il2.track_id FROM invoice_line il2
                WHERE il2.invoice_id = i_s.invoice_id
                ) IS NULL
            AND
               (
                SELECT il2.track_id FROM invoice_line il2
                WHERE il2.invoice_id = i_s.invoice_id
        
                EXCEPT
        
                SELECT t.track_id FROM track t
                WHERE t.album_id = (
                                    SELECT t2.album_id FROM track t2
                                    WHERE t2.track_id = i_s.last_track_id
                                   )
                ) IS NULL
        
            THEN "Album"
            ELSE "Tracks"
        END AS "purchase_type"
    FROM invoice_schema i_s
    )
GROUP BY 1
ORDER BY 2 DESC

Done.


[('Tracks', 500, 81.43), ('Album', 114, 18.57)]

## Conclusion

Entire album purchases are the 18.6% of the total purchases. That said, we recommend to continue to buy full albums from record companies otherwise there may be a potential loss of one fifth or so of the revenues. 